src/NUnitEngine/nunit.engine/AsyncTestEngineResult.cs

```diff
@@ -0,0 +1,87 @@
+﻿// ***********************************************************************
+// Copyright (c) 2011-2015 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Threading;
+
+namespace NUnit.Engine
+{
+    /// <summary>
+    /// Encapsulates an ongoing <see cref="TestEngineResult"/> 
+    /// </summary>
+    public class AsyncTestEngineResult
+    {
+        private readonly ManualResetEvent _waitHandle;
+        private volatile TestEngineResult _result;
```

Comment:
Probably not needed since we don't actually return the value until after the test is complete. It's probably not doing any harm, unless it causes extra work to update the result whenever we access it. AFAICS, we shouldn't be accessing it anyway until the test completes, at which point we can set it, right?